<a href="https://colab.research.google.com/github/michaelryanka/ASG-PYTN-H8-Batch-062/blob/main/PYTN_Assgn_3_Batch_062_Muhammad_Michael_Ryanka.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Pertama kita Import Packages, Functions, and Classes

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from sklearn.metrics import classification_report, confusion_matrix
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import accuracy_score



In [ ]:
df = pd.read_excel('bank.xlsx')
df
# kolom y adalah apakah client subscribed a term deposit? (binary: 'yes', 'no')

age            job  marital  education  ... pdays  previous poutcome   y
0      30     unemployed  married    primary  ...    -1         0  unknown  no
1      33       services  married  secondary  ...   339         4  failure  no
2      35     management   single   tertiary  ...   330         1  failure  no
3      30     management  married   tertiary  ...    -1         0  unknown  no
4      59    blue-collar  married  secondary  ...    -1         0  unknown  no
...   ...            ...      ...        ...  ...   ...       ...      ...  ..
4516   33       services  married  secondary  ...    -1         0  unknown  no
4517   57  self-employed  married   tertiary  ...    -1         0  unknown  no
4518   57     technician  married  secondary  ...    -1         0  unknown  no
4519   28    blue-collar  married  secondary  ...   211         3    other  no
4520   44   entrepreneur   single   tertiary  ...   249         7    other  no

[4521 rows x 17 columns]

In [ ]:
df.isnull().sum()

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4521 entries, 0 to 4520
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        4521 non-null   int64 
 1   job        4521 non-null   object
 2   marital    4521 non-null   object
 3   education  4521 non-null   object
 4   default    4521 non-null   object
 5   balance    4521 non-null   int64 
 6   housing    4521 non-null   object
 7   loan       4521 non-null   object
 8   contact    4521 non-null   object
 9   day        4521 non-null   int64 
 10  month      4521 non-null   object
 11  duration   4521 non-null   int64 
 12  campaign   4521 non-null   int64 
 13  pdays      4521 non-null   int64 
 14  previous   4521 non-null   int64 
 15  poutcome   4521 non-null   object
 16  y          4521 non-null   object
dtypes: int64(7), object(10)
memory usage: 600.6+ KB


In [ ]:
df.describe()

age       balance  ...        pdays     previous
count  4521.000000   4521.000000  ...  4521.000000  4521.000000
mean     41.170095   1422.657819  ...    39.766645     0.542579
std      10.576211   3009.638142  ...   100.121124     1.693562
min      19.000000  -3313.000000  ...    -1.000000     0.000000
25%      33.000000     69.000000  ...    -1.000000     0.000000
50%      39.000000    444.000000  ...    -1.000000     0.000000
75%      49.000000   1480.000000  ...    -1.000000     0.000000
max      87.000000  71188.000000  ...   871.000000    25.000000

[8 rows x 7 columns]

# **Data Prepration untuk Machine Learning**

Pertama kita melakukan Preprocessing terlebih dahulu dengan melakukan encoding

In [ ]:
#melakukan encode
encode = preprocessing.LabelEncoder()
df['job']=encode.fit_transform(df['job'])
df['marital']=encode.fit_transform(df['marital'])
df['education']=encode.fit_transform(df['education'])
df['default']=encode.fit_transform(df['default'])
df['housing']=encode.fit_transform(df['housing'])
df['loan']=encode.fit_transform(df['loan'])
df['contact']=encode.fit_transform(df['contact'])
df['month']=encode.fit_transform(df['month'])
df['poutcome']=encode.fit_transform(df['poutcome'])
df['y']=encode.fit_transform(df['y'])







Disini kita melakukan drop kepada kolom yang tidak dibutuhkan

In [ ]:
df.drop(['campaign','balance','duration'], axis=1, inplace=True) #tidak diperlukan
df

age  job  marital  education  ...  pdays  previous  poutcome  y
0      30   10        1          0  ...     -1         0         3  0
1      33    7        1          1  ...    339         4         0  0
2      35    4        2          2  ...    330         1         0  0
3      30    4        1          2  ...     -1         0         3  0
4      59    1        1          1  ...     -1         0         3  0
...   ...  ...      ...        ...  ...    ...       ...       ... ..
4516   33    7        1          1  ...     -1         0         3  0
4517   57    6        1          2  ...     -1         0         3  0
4518   57    9        1          1  ...     -1         0         3  0
4519   28    1        1          1  ...    211         3         1  0
4520   44    2        2          2  ...    249         7         1  0

[4521 rows x 14 columns]

In [ ]:
df.y.value_counts()
#no = 0
#yes = 1

0    4000
1     521
Name: y, dtype: int64

**Pertama kita menyiapkan features dan label nya dari data yang sudah di encoding, lalu kita dapat membuat dan menentukan model klasifikasi yang terdiri dari Logistic Regression, Decision Tree, Random Forest, SVM, Naive Bayes, KNN**

In [ ]:
features = ['age', 'job', 'marital', 'education', 'default', 'housing',
       'loan', 'contact', 'day', 'month', 'pdays',
       'previous', 'poutcome']
label = ['y']

In [ ]:
x = df[features]
y = df[label]


Train-validation-test split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.5, random_state=1) 

Implementasi Model

In [ ]:
#Logistic Regression
model=LogisticRegression()
model.fit(X_train,y_train)
pred_b=model.predict(X_test)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [ ]:
#Decision Trees
dtc = DecisionTreeClassifier()
dtc = dtc.fit(X_train,y_train)
pred_b1 = dtc.predict(X_test)

In [ ]:
#Random Forest
clf=RandomForestClassifier(n_estimators=100)
clf.fit(X_train,y_train) 
pred_b2=clf.predict(X_test)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
#SVM
from sklearn import svm
svm_model=svm.SVC()
svm_model.fit(X_train,y_train)
pred_b3=svm_model.predict(X_test)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
#Naive Bayes
from sklearn.naive_bayes import GaussianNB 
nb=GaussianNB()
nb.fit(X_train,y_train)
pred_b4=nb.predict(X_test)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
#KNN
from sklearn.neighbors import KNeighborsClassifier
kNN=KNeighborsClassifier()
kNN.fit(X_train,y_train)
pred_b5=kNN.predict(X_test)

/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [ ]:
#Hasil dari implementasi model
print("Logistic Regression:", accuracy_score(y_test,pred_b))
print("Decision Tree:", accuracy_score(y_test,pred_b1))
print("Random Forest:", accuracy_score(y_test,pred_b2))
print("SVM:", accuracy_score(y_test,pred_b3))
print("Naive Bayes:", accuracy_score(y_test,pred_b4))
print("KNN:", accuracy_score(y_test,pred_b5))

Logistic Regression: 0.8841220698805838
Decision Tree: 0.8142414860681114
Random Forest: 0.8911985846970367
SVM: 0.8850066342326405
Naive Bayes: 0.8111455108359134
KNN: 0.8779301194161875


**Dari hasil implementasi 6 model algoritma machine learning, bisa diketahui bahwa  Logistic Regression dan SVM mempunyai accuracy sebesar 88% dan Random Forest mempunyai accuracy 89%, maka dari itu menurut saya algoritma model machine learning yang paling sesuai dari hasil analisa adalah menggunakan algoritma Random Forest**